### ISSUE TITLE
Specify account confirmation flows

### ISSUE THREAD

**Issue by mikehedman** [hoodiehq/hoodie-account-server#2]
Email would probably be the default flow, but the specifics of this flow need to be specified, something like:
1. New user signs up
2. Account is put in unconfirmed state
3. Server sends mail to user
4. User clicks on link
5. Server could either confirm right then, or redirect the browser back to the client app and include the provided token, which the client could then submit using the account.confirm api.

Should also consider other possible verification options.  For example, on signup the server could send the user a text message with a code, the user then enters that code in the app, which would get submitted through the account.confirm api.
This SMS use case brings up a "feature" that should probably be included as we consider different confirmation paths: giving the app administrator a way to configure the nature of the token.  For email verification, it would likely be a somewhat long hash.  But for SMS, the admin might want the token/code to be just letters and no more than 4-8 characters.

Some financial applications do account verification through micro deposits to bank accounts, but that's pretty clearly out of scope for now :)

And yet another use case: I suppose an app should be able to "opt out" of confirming.

**Comment by gr2m** [comment]
I'd suggest we split up this issue, once we agreed on the different confirmation flows.

I'd suggest, we should have three confirmation workflows built-in, that can be defined with a `"confirm"` option, e.g.

```js
new HoodieAccountServer({
  confirm: 'auto'
})
```

The 3 built-in options would be 

1. "auto"
   That's what we have right now. A new user account gets confirmed automatically
2. "manual"  
   Admin has to sign in to the admin UI and confirm a user manually
3. "email"
   E-Mail confirmation workflow, as defined by Mike above.

Besides the built-in workflow, we should expose a way to define custom confirmation workflows, like the one via SMS or bank deposits. For that, the option could either be set to "manual" and an external service would listen to CouchDB's users and do its thing, or the "confirm" method could be set to a callback function, which get's called when ever a user signs up, does its work, and eventually confirms the user or sets a confirmation error.

Would that makes sense? We could split up this issue into 4 issues then, 3 for the built-in workflows, and one to discuss APIs and implementation for custom confirmation workflows

**Comment by mikehedman** [comment]
First, I agree that there will be a number of "go into the details" or "implement part x" issues that arise from this one.  But for now, I think it might be best to stay with just the one issue to decide a top level strategy.

As for the organization of options, I kind of like how it is now: by default the core will auto confirm, but plugins can register a callback to offer their opinion about whether the user should be confirmed.  So rather than building in the manual, invite-only, sms, or email functionality, I would offer them as optional plugins.  Or (and maybe it waits for a version 2) that could be extended by having the user object remember the "liens" against them.  So an app could include the manual and the email plugin, and at signup the plugins would register themselves in a user object attribute of "pending_confirmations": "{["manual", "email"]}".  I say this because I think many apps would want email no matter what, but then also might want the admin to decide.
The side benefit is that by implementing each flow as a plugin, we are also creating examples that others could extend, like to do SMS, or "watch 3 training videos", or "answer the new user survey", or "pay the dues."

**Comment by gr2m** [comment]
Current state:

1. Default will remain "auto-confirm"
2. `confirmation` option: https://github.com/hoodiehq/hapi-account/tree/master/server#optionsconfirmation can be set to `false`, `invite-only`, `"email"`. 
3. When set to "email", an email will be sent with a token that can be passed to `account.confirm(token)`
4. When set to `false`, only admins and 3rd-party plugins can confirm user accounts. The [account admin client](https://github.com/hoodiehq/hapi-account/tree/master/client/admin) will provide an according method
5. `invite-only` means that either admins or 3rd-party plugins can create user accounts that can be confirmed using `hoodie.account.signUp({username: '...', password: '...', token: '...'})`

**Comment by mikehedman** [comment]
A quick comment on item 3 in your last comment: I think the most commonly used use case is that the token is encoded in a url that the user clicks in the email.  Only if that fails would a user want to copy the token from the mail and paste it into a form on the site (which would presumably then submit it through account.confirm(token).

**Comment by gr2m** [comment]
Yeah, I should make it more clear. I'd definitely send a URL with the email, but the URL would ideally directly link to the app, which would read out the token and pass it to `account.confirm(token)`. So this would be app-specific logic, and the URL is configurable in `notifications.templates.confirmation`.

We could also add a URL to the REST API, which would directly confirm an account, and then redirect, or just show a message that the account is confirmed now. But when we link directly to the app, we can directly sign in the user, too.

**Comment by mikehedman** [comment]
Unless you know of some 'voodoo magic' for how to have a URL inside of an email route directly to an app running in a browser, I think we *must* have the URL point to a server side REST API and redirect.

**Comment by gr2m** [comment]
> 'voodoo magic' for how to have a URL inside of an email route directly to an app

I think there are ways, I'm sure apps do that already. But I'm not a native app developer. I think apps can register protocols, like `myapp://` or something? There is some "deep linking into apps" things from both Google and Apple, I just can't recall it right now

**Comment by mikehedman** [comment]
Ok, yes, one can do that - a native app can register itself to respond to a url.  But, there are a couple of missing pieces to the puzzle.  First, the server sending the confirm email wouldn't know what device would be used to read the mail and click on it. In a served web app, it could sense the user agent and show a page with an appropriate url.  But not so with email. There is also an assumption that the Hoodie app would be a native app - but while that may be a popular use, it would not be 100%.
So, for version 1, I would strongly lean towards just doing the 'standard' URL pointing to a REST API that redirects back to a browser.  From there, native apps can try to grab control (I think LinkedIn does this - but it starts in the browser first).
